xcm/xcm-executor/src/lib.rs

```diff
@@ -476,73 +503,94 @@ impl<Config: config::Config> XcmExecutor<Config> {
 		);
 		match instr {
 			WithdrawAsset(assets) => {
-				// Take `assets` from the origin account (on-chain) and place in holding.
-				let origin = *self.origin_ref().ok_or(XcmError::BadOrigin)?;
-				for asset in assets.into_inner().into_iter() {
-					Config::AssetTransactor::withdraw_asset(&asset, &origin, Some(&self.context))?;
-					self.subsume_asset(asset)?;
-				}
-				Ok(())
+				Config::TransactionalProcessor::process_transaction(|| -> Result<(), XcmError> {
```

Comment:
The naming is a bit weird because it's not actually processing a transaction, it's an instruction. Could this also be used once outside the `match instr` line rather than repeating it for each arm?

Reply:
Based on the [issue description](https://github.com/paritytech/polkadot/issues/6120) certain instructions are wrapped in the transactional processing.

Reply:
Doh, yeah OK. But what's not clear from the issue is whether the entire program should be transactional or just instructions? I.e. e.g. if the 3rd instruction has an error, would state changes from the first two (transactional) instructions also be rolled back?

Reply:
I don't think so or the whole Error handling logic of the XCVM wouldn't make sense anymore and people could execute XCM programs freely as the BuyExecution instruction would roll back as well.

Reply:
OK, still think `process_transaction` is confusing. Perhaps `TransactionalProcessor::process()`? The trait name indicates its behavior, the function says what it's doing. As in, would would expect `process` to process something transactionally within a transactional processor.